synapse/storage/databases/main/end_to_end_keys.py

```diff
@@ -367,6 +367,57 @@ def _count_e2e_one_time_keys(txn):
             "count_e2e_one_time_keys", _count_e2e_one_time_keys
         )
 
+    async def set_e2e_fallback_keys(
+        self, user_id: str, device_id: str, fallback_keys: dict
+    ) -> None:
+        """Set the user's e2e fallback keys.
+
+        Args:
+            user_id: the user whose keys are being set
+            device_id: the device whose keys are being set
+            fallback_keys: the keys to set.  This is a map from key ID (which is
+                of the form "algorithm:id") to key data.
+        """
+        # fallback_keys will usually only have one item in it, so using a for
+        # loop (as opposed to calling simple_upsert_many_txn) won't be too bad
+        # FIXME: make sure that only one key per algorithm is uploaded
```

Comment:
Do we need to fix this FIXME before merging?

Should this loop be in a transaction?

Reply:
I think that the FIXME is more of a "nice to have" rather than a requirement, so I don't think it *needs* to be fixed.

Also, I don't think the loop needs to be in a transaction (all the additions are independent), but I can do that if it you think it's a good idea.

Reply:
There's a constraint on the database for user + device + algorithm. Does that handle this or does this really only expect a single algorithm per user?

Reply:
It should handle it.  It would just do a series of upserts, so would just end up with the last fallback key and dropping all the others.  A client shouldn't expect that the server would store all the keys if it gives multiple keys per algorithm, so this seems like a reasonable result if we don't throw an error.

Reply:
So this fixme is really an optimization?